crates/controller/src/hud/minimap/fill.rs

```diff
@@ -1,27 +1,75 @@
+use super::draw::DrawingParam;
+use crate::ray::ScreenRay;
+use ahash::AHashMap;
 use bevy::{ecs::system::SystemParam, prelude::*};
 use de_core::{
-    gamestate::GameState, gconfig::GameConfig, objects::ObjectTypeComponent,
-    player::PlayerComponent, schedule::PostMovement,
+    gamestate::GameState, objects::ObjectTypeComponent, player::PlayerComponent,
+    schedule::PostMovement,
 };
 use de_map::size::MapBounds;
 use de_objects::SolidObjects;
 use de_terrain::TerrainCollider;
-use de_types::projection::ToFlat;
+use de_types::{objects::*, player::Player, projection::ToFlat};
 use parry2d::{
     bounding_volume::Aabb,
     math::Point,
     query::{Ray, RayCast},
 };
 
-use super::draw::DrawingParam;
-use crate::ray::ScreenRay;
-
 const TERRAIN_COLOR: Color = Color::rgb(0.61, 0.46, 0.32);
-const PLAYER_COLOR: Color = Color::rgb(0.1, 0.1, 0.9);
-const ENEMY_COLOR: Color = Color::rgb(0.9, 0.1, 0.1);
+const PLAYER_COLORS: [Color; Player::MAX_PLAYERS] = [
+    Color::rgb(0.1, 0.1, 0.9),
+    Color::rgb(0.1, 0.9, 0.1),
+    Color::rgb(0.9, 0.1, 0.1),
+    Color::rgb(0.9, 0.9, 0.1),
+];
 const MIN_ENTITY_SIZE: Vec2 = Vec2::splat(0.02);
 const CAMERA_COLOR: Color = Color::rgb(0.9, 0.9, 0.9);
 
+#[derive(Resource, Debug)]
+struct PlayerColors {
+    player_colors: [Color; Player::MAX_PLAYERS],
+    object_colors: [AHashMap<ObjectType, Color>; Player::MAX_PLAYERS],
+}
+impl PlayerColors {
+    fn new(colors: [Color; Player::MAX_PLAYERS]) -> Self {
+        Self {
+            player_colors: colors,
+            object_colors: [
+                derive_colors(colors[0]),
+                derive_colors(colors[1]),
+                derive_colors(colors[2]),
+                derive_colors(colors[3]),
+            ],
+        }
+    }
+
+    fn get_color(&self, player: Player, object_type: ObjectType) -> Color {
+        let player_idx: usize = (player.to_num() - 1) as usize;
+        let default_color: Color = self.player_colors[player_idx];
+        let object_color = self.object_colors[player_idx].get(&object_type);
```

Comment:
Instead of having a hash map, I would create a `match` here and multiply rgb values of default color with different factors in each match arm. That would be way faster. Example for inspiration:

```rust
let player_idx: usize = (player.to_num() - 1) as usize;
let player_color = self.player_colors[player_idx].as_hsla();
match object_type {
    ActiveObjectType::Building => player_color,
    ActiveObjectType::Unit(_) => player_color.with_s(0.7 * player_color.s())
}
```

Reply:
this implementation was much faster than mine and made the tuple struct work